src/query/service/src/interpreters/util.rs

```diff
@@ -61,3 +84,300 @@ pub fn generate_desc_schema(
+                    value: Literal::Decimal256 {
+                        value: (*v).into(),
+                        precision: MAX_DECIMAL256_PRECISION,
+                        scale: 0,
+                    },
+                }),
+                target_type: TypeName::Date,
+                pg_style: false,
+            },
+            Scalar::Timestamp(v) => Expr::Cast {
+                span: None,
+                expr: Box::new(Expr::Literal {
+                    span: None,
+                    value: Literal::Decimal256 {
+                        value: (*v).into(),
+                        precision: MAX_DECIMAL256_PRECISION,
+                        scale: 0,
+                    },
+                }),
+                target_type: TypeName::Timestamp,
+                pg_style: false,
+            },
+            Scalar::Null => Expr::Literal {
+                span: None,
+                value: Literal::Null,
+            },
+            Scalar::Bitmap(_) | Scalar::Binary(_) | Scalar::Geometry(_) | Scalar::Geography(_) => {
+                return Err(ErrorCode::Unimplemented(format!(
+                    "variable of type {} is not supported yet",
+                    scalar.as_ref().infer_data_type()
+                )));
+            }
+        };
+
+        Ok(ast)
+    }
+
+    fn read_from_set(
+        &self,
+        set: &Self::Set,
+        row: usize,
+        col: &ColumnAccess,
+    ) -> databend_common_exception::Result<Self::Var> {
+        let offset = match col {
+            ColumnAccess::Position(offset) => *offset,
+            // TODO(andylokandy): name resolution
+            ColumnAccess::Name(name) => set
+                .schema
+                .fields()
+                .iter()
+                .position(|f| f.name() == name)
+                .ok_or_else(|| {
+                ErrorCode::ScriptExecutionError(format!(
+                    "cannot find column with name {} in block, available columns: {}",
+                    name,
+                    set.schema
+                        .fields()
+                        .iter()
+                        .map(|f| format!("'{}'", f.name()))
+                        .join(", ")
+                ))
+            })?,
+        };
+        let col = set.block.columns().get(offset).ok_or_else(|| {
+            ErrorCode::ScriptExecutionError(format!(
+                "cannot read column at offset {} from block with {} columns",
+                offset,
+                set.block.num_columns()
+            ))
+        })?;
+        let cell = col
+            .value
+            .index(row)
+            .ok_or_else(|| {
+                ErrorCode::ScriptExecutionError(format!(
+                    "cannot read value at row {} from column with {} rows",
+                    row,
+                    set.block.num_rows(),
+                ))
+            })?
+            .to_owned();
+
+        Ok(cell)
+    }
+
+    fn set_len(&self, set: &Self::Set) -> usize {
```

Comment:
maybe better rename to `num_rows`. bad decision before XD

Reply:
Done